Scala 3: Add support to format open classes.

olafurpg opened the pull request:
This PR implements support to format open classes, which is a new soft
modifier in Scala 3. Details
https://dotty.epfl.ch/docs/reference/other-new-features/open-classes.html

tgodzik reviewed (approved):
LGTM! And we thought it would be a super quick feature :sweat_smile:

kitbellew commented on scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala:
```diff
@@ -87,10 +87,13 @@ object ScalafmtRunner {
-      .withAllowAndTypes(true) // New feature in Dotty
-      .withAllowTraitParameters(true) // New feature in Dotty
-      .withAllowColonForExtractorVarargs(true) // New feature in Dotty
+      // Supported Dotty features.
```
unrelated question: for the other changes that @MichalBednarz and @tgodzik have been working on (like `using` and `enum`), do we need to add these properties?

I can see one for `.withAllowEnums`, and something about `.withAllowGivenUsing` (although i am not sure if that's the same as `using`).

tgodzik replied:
I think we will use the separate Dotty dialect for now (until parser is fully stable) and require users to add `runner.dialect`. I think @olafurpg already included it in another PR for macros.

kitbellew commented on scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:
```diff
@@ -1,24 +1,34 @@
+      name: Name,
+      c: Classifier[Tree, T]
+  ): ModKey = ModKey(name.value, m => c(m))
+  val `private` = createMod[Mod.Private]
```
i'd add an empty line before this

kitbellew commented on scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:
```diff
@@ -1,24 +1,34 @@
+      matches: Mod => Boolean
+  ) extends Product
+  private def createMod[T <: Mod](implicit
+      name: Name,
```
also, question (on scala, probably): how does this work? why is there a unique implicit `Name` and `Classifier`?

poslegm replied:
there is [macro machinery](https://github.com/lihaoyi/sourcecode/blob/master/sourcecode/src-2/sourcecode/Macros.scala#L72-L77) around it

kitbellew commented on scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:
```diff
@@ -1,24 +1,34 @@
+  final case class ModKey(
+      name: String,
+      matches: Mod => Boolean
+  ) extends Product
```
a case class always extends `Product`, no?

olafurpg replied:
Correct, removed.

kitbellew commented on scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:
```diff
@@ -1,24 +1,34 @@
+      name: String,
+      matches: Mod => Boolean
+  ) extends Product
+  private def createMod[T <: Mod](implicit
```
i'd add an empty line before this

olafurpg replied:
Removed this method. Inserted a blank line after the class definition.

kitbellew commented on scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:
```diff
@@ -1,24 +1,34 @@
+  private def createMod[T <: Mod](implicit
+      name: Name,
+      c: Classifier[Tree, T]
+  ): ModKey = ModKey(name.value, m => c(m))
```
why not `c.apply`?

olafurpg replied:
No specific reason. I admit this method was unnecessarily complicated so I removed the method.

kitbellew reviewed (changes requested):
there's some scala magic happening here :) but a few questions still...

olafurpg commented:
Superseded by #2286, where this commit https://github.com/scalameta/scalafmt/pull/2286/commits/6e0878fb8654e3cb64a97655e91bed3c032cfb0c addresses the comments from this PR
